---
title: fastqc-viz
title-block-banner: false

format:
  html:
    mainfont: "DM Sans"
   
execute: 
  warning: false
  message: false
  echo: false

sidebar: false
---

```{r}
#| label: load functions
library(tidyverse)
purrr::walk(list.files("R", pattern = "\\.R$", full.names = TRUE), source)
theme_set_fastqcviz()
```

```{r}
#| label: load and parse fasqc data
fastqc_data <-
  parse_fastqc(fastq_data_file = "output/fastqc_reports/SRR1517848_2_fastqc.txt")
```

::::::: grid-container
::: {.box .sidebox}

![](images/logo-light.svg){width=100%, id="logo"}

-------------

```{r}
#| label: Status summary
#| results: asis
plot_status_summary(fastqc_data) 

```

```{r}
#| label: Basic Statistics [title]
#| results: asis

create_header(fastqc_data, "basic_statistics")
```

```{r}
#| label: Basic Statistics
#| results: asis
plot_basic_statistics(fastqc_data)
```

:::

::: box
```{r}
#| label: Per base sequence quality [title]
#| results: asis
create_header(fastqc_data, "per_base_sequence_quality")
```

```{r}
#| label: Per base sequence quality [content]
#| fig-width: 3
#| fig-height: 1.75


fqcviz_colors <- get_color_palette()

data2plot <-
  fastqc_data$per_base_sequence_quality$content |>
  separate(base, into = c('start', 'end'), sep = '-') |>
  mutate(across(c(start, end), as.numeric)) |>
  rowwise() |>
  mutate(base_numeric = mean(c(start, end), na.rm = TRUE))

p1 <- 
  data2plot |>
  ggplot(aes(x = base_numeric, y = as.numeric(mean))) +
  # Fail zone
  annotate(
    geom = 'rect',
    xmin = -Inf,
    xmax = Inf,
    ymin = -Inf,
    ymax = 20,
    fill = fqcviz_colors$warm_grey3,
    alpha = .4
  ) +
  # Warn zone
  annotate(
    geom = 'rect',
    xmin = -Inf,
    xmax = Inf,
    ymin = 20,
    ymax = 28,
    fill = fqcviz_colors$warm_grey4,
    alpha = .4
  ) +
  # Pass zone
  annotate(
    geom = 'rect',
    xmin = -Inf,
    xmax = Inf,
    ymin = 28,
    ymax = Inf,
    fill = fqcviz_colors$warm_grey5,
    alpha = .4
  ) +
  
  # Zone delimiter
  # fail-warn limit
  # annotate(
  #   geom = 'segment',
  #   x = -Inf,
  #   xend = Inf,
  #   y = 20,
  #   yend = 20,
  #   color = "white"
  # ) +
  # # warn-pass limit
  # annotate(
  #   geom = 'segment',
  #   x = -Inf,
  #   xend = Inf,
  #   y = 28,
  #   yend = 28,
  #   color = "white"
  # ) +
  
  # main geometries
  # boxplot whiskers
  geom_errorbar(
    aes(
      x = base_numeric,
      ymin = as.numeric(x10th_percentile),
      ymax = as.numeric(x90th_percentile)
    ),
    color = fqcviz_colors$warm_grey2,
    linewidth = .5,
    width = .6,
    alpha = .2
  ) +
  # boxpot box
  geom_segment(
    aes(
      x = base_numeric,
      y = as.numeric(upper_quartile),
      yend = as.numeric(lower_quartile)
    ),
    color = fqcviz_colors$warm_grey2,
    linewidth = 2,
    alpha = .2
  ) +
  # boxplot median
  geom_errorbar(
    aes(
      x = base_numeric,
      ymin = as.numeric(median),
      ymax = as.numeric(median)
    ),
    color = fqcviz_colors$warm_grey2,
    size = .5,
    width = 1,
    alpha = .1
  ) +
  # main distribuition
  geom_point(size = .5, color = fqcviz_colors$blue1) +
  geom_line(aes(group = 1), color = fqcviz_colors$blue1) +
  scale_y_continuous(
    limits = c(0, max(as.numeric(data2plot$upper_quartile)))
  ) +
  scale_x_continuous(
    expand = c(0, 0),
    limits = c(0, max(as.numeric(data2plot$base_numeric)) + 1),
    breaks = c(min(data2plot$base_numeric), max(data2plot$base_numeric))
  ) +
  labs(
    x = "Read base position",
    y = "Quality"
  )


p1

```

:::

::: box
```{r}
#| label: Per base sequence content [title]
#| results: asis
create_header(fastqc_data, "per_sequence_quality_scores")
```

```{r}
#| label: Per base sequence content [content]
#| fig-width: 3
#| fig-height: 1.75


fqcviz_colors <- get_color_palette()

data2plot <-
  fastqc_data$per_sequence_quality_scores$content |>
  mutate(
    quality = as.numeric(quality),
    count = as.numeric(count))

# p1 <- 
data2plot |>
  ggplot(aes(x = quality, y = count)) +
  geom_point(size = .5, color = fqcviz_colors$blue1) +
  geom_line(aes(group = 1), color = fqcviz_colors$blue1) +
  scale_y_continuous(labels = format_large_numbers) +
  labs(
    x = "Average Quality per Read (Phred score)",
    y = "Number of Reads")



```

:::

::: box
```{r}
#| label: per_base_sequence_content [title]
#| results: asis
create_header(fastqc_data, "per_base_sequence_content")
```

```{r}
#| label: per_base_sequence_content [content]
#| fig-width: 3
#| fig-height: 1.75

qcviz_colors <- get_color_palette()

data2plot <-
  fastqc_data$per_base_sequence_content$content |>
  pivot_longer(
    cols = c('a', 'c', 'g', 't'),
    names_to = "nucleotide",
    values_to = "count"
  ) |>
  mutate(
    count = as.numeric(count)
  ) |>
  separate(base, into = c('start', 'end'), sep = '-') |>
  mutate(across(c(start, end), as.numeric)) |>
  rowwise() |>
  mutate(base_numeric = mean(c(start, end), na.rm = TRUE))

data2plot |>
  ggplot(aes(x = base_numeric, y = count)) +
  geom_line(aes(color = nucleotide), linewidth = .5) +
  scale_y_continuous(
    labels = format_large_numbers,
    limits = c(0, 100),
    expand = c(0, 0)
  ) +
  scale_x_continuous(
    limits = c(0, max(data2plot$base_numeric)),
    breaks = c(min(data2plot$base_numeric) - 1, max(data2plot$base_numeric)),
    expand = c(0, 0)
  ) +
  labs(
    x = "Position in read (bp)",
    y = "Nucleotide Fraction (%)",
    color = NULL
  ) +
  ggsci::scale_color_lancet(label = toupper) +
  theme(
    legend.position = c(0.5, 1),
    legend.justification = c(0.5, 1),
    legend.direction = "horizontal"
  )
```

:::

::: box
```{r}
#| label: Per sequence GC content [title]
#| results: asis
create_header(fastqc_data, "per_sequence_gc_content")
```

```{r}
#| label: Per sequence GC content [content]
#| fig-width: 3
#| fig-height: 1.75

qcviz_colors <- get_color_palette()

data2plot <-
  fastqc_data$per_sequence_gc_content$content |>
  mutate(
    gc_content = as.numeric(gc_content),
    count = as.numeric(count)
  )

theoretical_data <-
  process_gc_data(data2plot)

data2plot <-
  data2plot |>
  mutate(group = 'Observed') |>
  bind_rows(
    theoretical_data$normal_distribution_df |>
      mutate(group = 'Theoretical')
  ) |>
  mutate(group = factor(group, levels = c('Theoretical', 'Observed')))

data2plot |>
  ggplot(aes(x = gc_content, y = count)) +
  geom_line(aes(color = group), linewidth = .5) +
  labs(
    x = "Mean GC content (%)",
    y = "Count (n)",
    color = NULL
  ) +
  scale_y_continuous(labels = format_large_numbers) +
  scale_color_manual(
    values = c(
      "Observed" = fqcviz_colors$blue1,
      "Theoretical" = fqcviz_colors$warm_grey4
    )
  ) +
  theme(legend.position = c(1, 1), legend.justification = c(1, 1),legend.key.spacing.y = unit(.1, "cm"), legend.key.height = unit(.1, "cm"))


```

:::


<!-- End grid container -->
:::::::